torch/autograd/functional.py

```diff
@@ -0,0 +1,714 @@
+                                   " the hessian depends on the inputs.".format(i))
+            elif input_type == "jacobian":
+                raise RuntimeError("The jacobian of the user-provided function with respect to input {}"
+                                   " is independent of the input. This is not allowed in strict mode."
+                                   " You should ensure that your function is twice differentiable and that"
+                                   " the jacobian depends on the inputs (this would be violated by a linear"
+                                   " function).".format(i))
+            elif input_type == "grad_inputs":
+                raise RuntimeError("The gradient with respect to input {} is independent of the inputs of the"
+                                   " user-provided function. This is not allowed in strict mode.".format(i))
+            else:
+                raise RuntimeError("Output {} of the user-provided function does not require gradients."
+                                   " The outputs must be computed in a differentiable manner from the input"
+                                   " when running in strict mode.".format(i))
+
+def _autograd_grad(outputs, inputs, grad_outputs=None, create_graph=False, retain_graph=None):
+    # Version of autograd.grad that accepts `None` in outputs and do not compute gradients for them.
+    # This has the extra constraint that inputs has to be a tuple
+    if torch.is_tensor(outputs):
+        outputs = (outputs,)
+        grad_outputs = (grad_outputs,)
+    if grad_outputs is None:
+        grad_outputs = (None,) * len(outputs)
+
+    new_outputs = tuple()
+    new_grad_outputs = tuple()
+    for out, grad_out in zip(outputs, grad_outputs):
+        if out is not None and out.requires_grad:
+            new_outputs += (out,)
+            new_grad_outputs += (grad_out,)
+
+    if len(new_outputs) == 0:
+        # No differentiable output, we don't need to call the autograd engine
+        return (None,) * len(inputs)
+    else:
+        return torch.autograd.grad(new_outputs, inputs, new_grad_outputs, allow_unused=True,
+                                   create_graph=create_graph, retain_graph=retain_graph)
+
+# Public API
+
+def vjp(func, inputs, v=None, create_graph=False, strict=False):
+    r"""Function that computes the dot product between a vector ``v`` and the Jacobian of
+    the given function at the point given by the inputs.
+
+    Args:
+        func (function): a Python function that takes Tensor inputs and returns
+            a tuple of Tensor or a Tensor.
+        inputs (tuple of Tensor or Tensor): inputs to the function.
+        v (tuple of Tensor or Tensor): The vector that will multiply the Jacobian. Must be the
+            same size as the output of ``func``. This argument is optional when
+            ``func``'s output contains a single element and it will be set as a Tensor
```

Comment:
nit: "This argument is optional when ``func``'s output contains a single element and (if it is not provided) it will be set as a Tensor"

Things in the parenthesis to make it unambiguous